Add alias reference support (`alias.<name>`)

EtgarDev opened the pull request:
### Summary of Changes
- Added support for `alias.name` in SELECT projection:
  - Introduced `SupportsAliasReference` and `TryResolveAliasReference` in `ExpressionBinder`.
  - Implemented `TryResolveAliasReference` specifically in `SelectBinder`.
  - Included robust tests for both valid aliases and edge cases.

Related issue: https://github.com/duckdb/duckdb/issues/13991#issuecomment-2357781600

<img width="281" height="229" alt="image" src="https://github.com/user-attachments/assets/31b673a3-0835-4618-add3-e4196abc1e2f" />

Works with:
- `SELECT`
- `GROUP BY` (works also with inside-expression aliases, `GROUP BY alias.x + 1`, which was impossible in legacy aliasing)
- `WHERE` / `HAVING` / `QUALIFY`
- `ORDER BY`

EtgarDev committed:
Introduce `alias_ref('name')` support in SELECT projection

- Added `SupportsAliasReference` and `TryResolveAliasReference` in `ExpressionBinder`.
- Implemented `TryResolveAliasReference` in `SelectBinder`.
- Created tests for alias_ref functionality, including both valid cases and error scenarios.

EtgarDev committed:
Expand `alias_ref` error handling tests for invalid and NULL arguments

EtgarDev committed:
Remove verification pragma from alias_function_errors test

EtgarDev committed:
prettier

EtgarDev committed:
removed redundant comments

EtgarDev committed:
removed redundant comments

Mytherin commented:
Thanks for the PR! Could we try to use a table qualification `alias` for this instead of a function as proposed in the linked issue? e.g.:

```sql
SELECT a + 1 AS a, alias.a
FROM (SELECT 42 a) t;
```

Instead of:

```sql
SELECT a + 1 AS a, alias_ref('a')
FROM (SELECT 42 a) t;
```

I think this might also simplify the implementation as `BindColumnRef` is already where aliases are bound normally.

EtgarDev committed:
Expand `alias_ref` error handling tests for subqueries in SELECT aliases

EtgarDev committed:
Refactor `alias_ref` to use `alias.name` syntax in SELECT projection

- Replaced `alias_ref('name')` usage with `alias.name`.
- Updated `SupportsAliasReference` and `TryResolveAliasReference` to handle `ColumnRefExpression` instead of `FunctionExpression`.
- Removed `alias_ref` binding logic from function expressions.
- Enhanced error handling for `alias.*` references in SELECT lists.
- Updated tests to reflect the new syntax.

EtgarDev committed:
Refactor `alias_ref` tests and binder logic to adopt `alias.name` syntax

- Replaced all instances of `alias_ref('name')` with `alias.name` in tests and binder implementation.
- Updated error messages and added stricter validation for `alias.*` usage outside SELECT projection lists.
- Enhanced parser and binder to handle `alias.name` syntax with improved error handling.

EtgarDev committed:
Rename `alias_function` tests to `alias_qualification` for clarity.

EtgarDev committed:
Expand `alias_qualification` tests to cover various alias usage scenarios

- Added tests for chained alias references, multiple alias references, mixed-case quoted aliases, and aliases with window functions.
- Included scenarios with DISTINCT, CASE expressions, and string aliases.
- Ensured deterministic ordering and validation in test cases.

Mytherin commented on src/planner/expression_binder/select_binder.cpp:
```diff
@@ -8,6 +9,33 @@ SelectBinder::SelectBinder(Binder &binder, ClientContext &context, BoundSelectNo
+unique_ptr<Expression> SelectBinder::TryResolveAliasReference(const string &alias_name,
+                                                              const ColumnRefExpression &column_ref_expr) {
+	// resolve alias.name within SELECT list
+	auto entry = node.bind_state.alias_map.find(alias_name);
```
There's already code for binding aliases in several places, e.g.:

* [Select](https://github.com/duckdb/duckdb/blob/58781d719cc5b020d3bba31a48526b0289e3a73a/src/planner/expression_binder/select_binder.cpp#L29)
* [Column Alias Binder](https://github.com/duckdb/duckdb/blob/58781d719cc5b020d3bba31a48526b0289e3a73a/src/planner/expression_binder/column_alias_binder.cpp#L20)
* [TryBindAlias](https://github.com/duckdb/duckdb/blob/58781d719cc5b020d3bba31a48526b0289e3a73a/src/planner/binder/expression/bind_columnref_expression.cpp#L516)

Rather than adding another path where aliases can be bound - I think it might be easier / more robust to change `!col_ref_p.IsQualified()` to e.g. `IsPotentialAlias(col_ref)` in those places - where that is something like:`

```cpp
bool ExpressionBinder::IsPotentialAlias(const ColumnRefExpression &colref) {
    if (colref.IsQualified()) {
         return true;
    }
    if (colref.column_names.size() == 2 && StringUtil::CIEquals(col_ref_p.GetTableName(), "alias")) {
        return true;
    }
    return false;
}
```


That way this can also be used in different locations in the query.

EtgarDev replied:
@Mytherin 
*This also reduces some redundant alias fallback work:
alias resolution no longer happens both inside `bind_columnref_expression` and again inside `SelectBinder` as a late fallback.
The new unified flow prevents double binding calls and avoids several redundant calls.
I'll try running a benchmark over it to see if there is a real improvement for queries with a lot of aliases.

Mytherin reviewed (commented):
Thanks for the fixes! Looks good - some more comments from my side

EtgarDev committed:
Fix conflicts with the edge-case of having a table named "alias"

EtgarDev commented on test/sql/binder/alias_qualification_select_projection.test:
```diff
@@ -0,0 +1,86 @@
+# description: Test alias.name resolution in SELECT projection
+# group: [binder]
+
+query II
```
Nice, fixed! :)

EtgarDev committed:
Refactor alias binding to implement unified two-step alias resolution

- Introduced `TryBindRegularAlias` and `TryResolveAliasReference` for distinct alias resolution steps (unqualified and qualified).
- Replaced `TryBindAlias` with the new methods across all binder contexts.
- Deprecated `SupportsAliasReference` in favor of explicit binding functions.
- Simplified alias detection with `ExpressionBinder::IsPotentialAlias`.

EtgarDev committed:
Add tests for alias resolution in various SQL clauses

- Introduced `alias_qualification_*` tests for GROUP BY, HAVING, ORDER BY, QUALIFY, and WHERE clauses.
- Covered scenarios for regular aliases, chaining aliases, quoted aliases, and handling tables named `alias`.
- Ensured consistent behavior and validation for `alias.name` syntax across all cases.

EtgarDev committed:
Simplify comment for legacy alias binding function

EtgarDev commented on src/planner/expression_binder/group_binder.cpp:
```diff
@@ -102,13 +109,41 @@ bool GroupBinder::TryBindAlias(ColumnRefExpression &colref, bool root_expression
 	// first try to bind to the base columns (original tables)
+
+	// mark that we are binding the root GROUP BY expression so alias rules apply
+	struct RootGuard {
```
To preserve old restriction for inside-expression aliases in legacy aliases

EtgarDev committed:
Simplify comment in `OrderBinder::BindExpression` to improve clarity

EtgarDev committed:
Remove invalid test cases for `alias.*` usage outside SELECT lists

EtgarDev committed:
Bringing back older functionality of `bind_columnref_expression` that I missed in the new implementation

EtgarDev committed:
formatter

Mytherin commented on src/planner/expression_binder/order_binder.cpp:
```diff
@@ -76,12 +76,21 @@ optional_idx OrderBinder::TryGetProjectionReference(ParsedExpression &expr) cons
+				break;
+			}
+		} else {
+			alias_name = colref.column_names[0];
```
Can we change this to use the unified codepath:

```cpp
if (!IsPotentialAlias(colref)) {
    // not an alias
    break;
}
string alias_name = colref.column_names.back();
```

EtgarDev replied:
nice catch, fixed! :)

Mytherin commented on src/planner/expression_binder/column_alias_binder.cpp:
```diff
@@ -16,13 +16,25 @@ bool ColumnAliasBinder::BindAlias(ExpressionBinder &enclosing_binder, unique_ptr
 	}
 
+	string alias_name;
+	if (expr.IsQualified()) {
```
`auto alias_name = expr.column_names.back();`

Mytherin reviewed (commented):
Thanks! Looks good - I think unifying and cleaning up alias binding is a great idea. I've left some more comments

EtgarDev committed:
Fixed resolving of the AliasReference after mistakenly removing it

EtgarDev committed:
Fixed a test to expect the right amount of columns

EtgarDev commented on src/planner/expression_binder/select_binder.cpp:
```diff
@@ -8,24 +9,7 @@ SelectBinder::SelectBinder(Binder &binder, ClientContext &context, BoundSelectNo
-	// binding failed
-	// check in the alias map
-	auto &colref = (expr_ptr.get())->Cast<ColumnRefExpression>();
+bool SelectBinder::TryBindRegularAlias(ColumnRefExpression &colref, BindResult &result) {
```
@Mytherin 
I actually struggled with this myself for a while, and ended up splitting the logic for a reason.

Unqualified aliases follow the SQL standard rules, they only exist in very specific scopes, can’t appear inside subqueries, must be defined earlier in the `SELECT` list.
The new `alias.name` form behaves differently on purpose - it acts like a namespace reference and can appear in more places, but still needs its own safety rules (in example - no subquery aliases, no circular referencing).

Because the two forms obey different scoping / visibility rules, I kept `TryBindRegularAlias` and `TryResolveAliasReference` separate, to preserve legacy SQL behavior exactly as before while adding the new syntax in a clean and predictable way.

What do you think?
Would you still suggest to unify them? Maybe a better naming would clarify the ambiguity?

EtgarDev committed:
Remove the redundent `BindColumnRef` function in `SelectBinder`

EtgarDev committed:
Refactor alias binding checks for consistency and simplicity

- Replaced repetitive alias qualification logic with `ExpressionBinder::IsPotentialAlias`.
- Unified alias name retrieval using `column_names.back()`.

EtgarDev committed:
Refactor alias binding in `OrderBinder` for clarity and consistency

- Replaced verbose alias qualification logic with `ExpressionBinder::IsPotentialAlias`.
- Simplified alias name retrieval using `column_names.back()`.

EtgarDev committed:
Remove redundant `BindColumnRef` override in `SelectBinder`

EtgarDev committed:
Add `depth` parameter to alias binding functions to fix errors in aliasing thorugh a subquery

- Updated alias-related methods to include `idx_t depth`.
- Refactored `TryBindRegularAlias` and `TryResolveAliasReference` to use `depth` for enhanced binding logic.

EtgarDev committed:
Remove `QualifyColumnAlias` method and refactor related alias binding logic

- Eliminated the `QualifyColumnAlias` method across binders.
- Refactored alias qualification to use `TryBindRegularAlias` consistently.
- Simplified alias handling and removed redundant code.

EtgarDev commented on src/include/duckdb/planner/expression_binder/where_binder.hpp:
```diff
@@ -24,7 +24,6 @@ class WhereBinder : public ExpressionBinder {
 	                          bool root_expression = false) override;
 
 	string UnsupportedAggregateMessage() override;
-	bool QualifyColumnAlias(const ColumnRefExpression &colref) override;
```
@Mytherin 
I managed to eliminate `QualifyColumnAlias` as it does the same as `TryBindRegularAlias` (which will be renamed soon after the unification we talked about)

EtgarDev committed:
Remove `depth` parameter from `BindAndQualifyFunction` and update calls accordingly

- Simplified `BindAndQualifyFunction` by removing unused `depth` parameter.
https://github.com/duckdb/duckdb/pull/19747#discussion_r2534467122

EtgarDev committed:
Remove `TryBindRegularAlias` and unify alias binding through `TryResolveAliasReference`

EtgarDev committed:
- Refactor alias binding by replacing `QualifyColumnAlias` with `DoesColumnAliasExist` for consistency and clarity across binders
- Use `DoesColumnAliasExist` where `QualifyColumnAlias` was used

EtgarDev committed:
Clean up: remove redundant blank lines and adjust indentation in expression binder files

EtgarDev committed:
Remove `DoesColumnAliasExist` from `GroupBinder`

EtgarDev committed:
Add test for method call on alias reference in projection

EtgarDev commented on src/include/duckdb/planner/expression_binder/where_binder.hpp:
```diff
@@ -24,7 +24,8 @@ class WhereBinder : public ExpressionBinder {
 	string UnsupportedAggregateMessage() override;
-	bool QualifyColumnAlias(const ColumnRefExpression &colref) override;
+
+	bool DoesColumnAliasExist(const ColumnRefExpression &colref) override;
```
Renamed for clarity

EtgarDev committed:
Remove unused `BindResult` and related comment in alias qualification logic

EtgarDev committed:
Add `root_expression` parameter to `TryResolveAliasReference` to eliminate the RAII guard of GroupBinder::BindColumnRef

EtgarDev committed:
Remove redundant comment in alias binding logic detailing the two-step process

EtgarDev committed:
Move `UnsupportedAggregateMessage` method implementation to reduce "changes" in PR

EtgarDev committed:
Remove unused `in_root_group_ref` variable from `GroupBinder`

EtgarDev committed:
Reformat

EtgarDev committed:
Replace exception with `false` return in alias binding when alias doesn't exist

EtgarDev committed:
Update tests to verify proper error handling for invalid alias usage in GROUP BY

EtgarDev committed:
Update test to reflect changes in alias qualification logic for GROUP BY and HAVING

EtgarDev committed:
The error is actually regarding unknown table but it doesn't matter

EtgarDev committed:
Update error messages for alias qualification logic in SELECT clause

EtgarDev committed:
- Fix WhereBinder to align with other binders
- Needed to add the original `expr_ptr` as a parameter to the signature of `bind_columnref_expression::BindExpression` to do so

EtgarDev committed:
Reformat

Mytherin reviewed (commented):
Thanks! LGTM - one minor comment then this is good to go

EtgarDev committed:
Fix alias reference in SELECT clause error message

EtgarDev commented on test/sql/binder/alias_qualification_errors.test:
```diff
@@ -0,0 +1,27 @@
+statement error
+SELECT alias.x + 1 AS y, 1 AS x
+----
+Binder Error: Column "alias" referenced that exists in the SELECT clause - but this column cannot be referenced before it is defined
```
totally right! I missed that as I copied & pasted the same error message from the previous implementation :)

EtgarDev committed:
Update error message for alias reference in SELECT clause test

EtgarDev commented:
> Thanks! LGTM - one minor comment then this is good to go

Thank you @Mytherin :)
Everything is ready and set now :D
